// @strapi/utils/dist/sanitize/sanitizers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultSanitizePopulate = exports.defaultSanitizeFields = exports.defaultSanitizeSort = exports.defaultSanitizeFilters = exports.defaultSanitizeOutput = exports.sanitizePasswords = void 0;
const fp_1 = require("lodash/fp");
const async_1 = require("../async");
const traverse_entity_1 = __importDefault(require("../traverse-entity"));
const content_types_1 = require("../content-types");
const traverse_1 = require("../traverse");
const visitors_1 = require("./visitors");
const operators_1 = require("../operators");
const sanitizePasswords = (schema) => async (entity) => {
    if (!schema) {
        throw new Error('Missing schema in sanitizePasswords');
    }
    return (0, traverse_entity_1.default)(visitors_1.removePassword, { schema }, entity);
};
exports.sanitizePasswords = sanitizePasswords;
const defaultSanitizeOutput = async (schema, entity) => {
    if (!schema) {
        throw new Error('Missing schema in defaultSanitizeOutput');
    }
    return (0, traverse_entity_1.default)((...args) => {
        (0, visitors_1.removePassword)(...args);
        (0, visitors_1.removePrivate)(...args);
    }, { schema }, entity);
};
exports.defaultSanitizeOutput = defaultSanitizeOutput;
const defaultSanitizeFilters = (0, fp_1.curry)((schema, filters) => {
    if (!schema) {
        throw new Error('Missing schema in defaultSanitizeFilters');
    }
    return (0, async_1.pipeAsync)(
    // Remove keys that are not attributes or valid operators
    (0, traverse_1.traverseQueryFilters)(({ key, attribute }, { remove }) => {
        const isAttribute = !!attribute;
        // ID is not an attribute per se, so we need to make
        // an extra check to ensure we're not checking it
        if (key === 'id') {
            return;
        }
        if (!isAttribute && !(0, operators_1.isOperator)(key)) {
            remove(key);
        }
    }, { schema }), 
    // Remove dynamic zones from filters
    (0, traverse_1.traverseQueryFilters)(visitors_1.removeDynamicZones, { schema }), 
    // Remove morpTo relations from filters
    (0, traverse_1.traverseQueryFilters)(visitors_1.removeMorphToRelations, { schema }), 
    // Remove passwords from filters
    (0, traverse_1.traverseQueryFilters)(visitors_1.removePassword, { schema }), 
    // Remove private from filters
    (0, traverse_1.traverseQueryFilters)(visitors_1.removePrivate, { schema }), 
    // Remove empty objects
    (0, traverse_1.traverseQueryFilters)(({ key, value }, { remove }) => {
        if ((0, fp_1.isObject)(value) && (0, fp_1.isEmpty)(value)) {
            remove(key);
        }
    }, { schema }))(filters);
});
exports.defaultSanitizeFilters = defaultSanitizeFilters;
const defaultSanitizeSort = (0, fp_1.curry)((schema, sort) => {
    if (!schema) {
        throw new Error('Missing schema in defaultSanitizeSort');
    }
    return (0, async_1.pipeAsync)(
    // Remove non attribute keys
    (0, traverse_1.traverseQuerySort)(({ key, attribute }, { remove }) => {
        // ID is not an attribute per se, so we need to make
        // an extra check to ensure we're not checking it
        if (key === 'id') {
            return;
        }
        if (!attribute) {
            remove(key);
        }
    }, { schema }), 
    // Remove dynamic zones from sort
    (0, traverse_1.traverseQuerySort)(visitors_1.removeDynamicZones, { schema }), 
    // Remove morpTo relations from sort
    (0, traverse_1.traverseQuerySort)(visitors_1.removeMorphToRelations, { schema }), 
    // Remove private from sort
    (0, traverse_1.traverseQuerySort)(visitors_1.removePrivate, { schema }), 
    // Remove passwords from filters
    (0, traverse_1.traverseQuerySort)(visitors_1.removePassword, { schema }), 
    // Remove keys for empty non-scalar values
    (0, traverse_1.traverseQuerySort)(({ key, attribute, value }, { remove }) => {
        // ID is not an attribute per se, so we need to make
        // an extra check to ensure we're not removing it
        if (key === 'id') {
            return;
        }
        if (!(0, content_types_1.isScalarAttribute)(attribute) && (0, fp_1.isEmpty)(value)) {
            remove(key);
        }
    }, { schema }))(sort);
});
exports.defaultSanitizeSort = defaultSanitizeSort;
const defaultSanitizeFields = (0, fp_1.curry)((schema, fields) => {
    if (!schema) {
        throw new Error('Missing schema in defaultSanitizeFields');
    }
    return (0, async_1.pipeAsync)(
    // Only keep scalar attributes
    (0, traverse_1.traverseQueryFields)(({ key, attribute }, { remove }) => {
        // ID is not an attribute per se, so we need to make
        // an extra check to ensure we're not checking it
        if (key === 'id') {
            return;
        }
        if ((0, fp_1.isNil)(attribute) || !(0, content_types_1.isScalarAttribute)(attribute)) {
            remove(key);
        }
    }, { schema }), 
    // Remove private fields
    (0, traverse_1.traverseQueryFields)(visitors_1.removePrivate, { schema }), 
    // Remove password fields
    (0, traverse_1.traverseQueryFields)(visitors_1.removePassword, { schema }), 
    // Remove nil values from fields array
    (value) => ((0, fp_1.isArray)(value) ? value.filter((field) => !(0, fp_1.isNil)(field)) : value))(fields);
});
exports.defaultSanitizeFields = defaultSanitizeFields;
const defaultSanitizePopulate = (0, fp_1.curry)((schema, populate) => {
    if (!schema) {
        throw new Error('Missing schema in defaultSanitizePopulate');
    }
    return (0, async_1.pipeAsync)((0, traverse_1.traverseQueryPopulate)(async ({ key, value, schema, attribute }, { set }) => {
        if (attribute) {
            return;
        }
        if (key === 'sort') {
            set(key, await defaultSanitizeSort(schema, value));
        }
        if (key === 'filters') {
            set(key, await defaultSanitizeFilters(schema, value));
        }
        if (key === 'fields') {
            set(key, await defaultSanitizeFields(schema, value));
        }
    }, { schema }), 
    // Remove private fields
    (0, traverse_1.traverseQueryPopulate)(visitors_1.removePrivate, { schema }))(populate);
});
exports.defaultSanitizePopulate = defaultSanitizePopulate;
